File: app/src/main/java/humbleactivity/app/ui/ChainComposer.java
package humbleactivity.app.ui;

import com.jakewharton.rxrelay.PublishRelay;
import humbleactivity.app.RxScheduling;
import humbleactivity.app.data.EffectorService;
import humbleactivity.app.data.Filter;
import rx.Observable;
import timber.log.Timber;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;

public class ChainComposer extends Presenter<ChainComposer.ChainComposerView> {
    private final EffectorService effectorService;
    private final RxScheduling rxScheduling;
    private final PublishRelay<Void> refreshRelay = PublishRelay.create();
    public List<Filter> availableFilters;
    public List<Filter> chain;

    @Inject
    public ChainComposer(EffectorService effectorService, RxScheduling rxScheduling) {
        this.effectorService = effectorService;
        this.rxScheduling = rxScheduling;
        Timber.d("Created");
    }

    public void attach(ChainComposerView view) {
        super.attach(view);
        subscriptions.add(refreshRelay.flatMap(none ->
                rxScheduling.ioThenUi(effectorService.listFilters())
                        .doOnError(throwable -> {
                            view.showErrorMessage(throwable.getMessage());
                        })
                        .onErrorResumeNext(Observable.empty())
        ).subscribe(filters -> {
            availableFilters = new ArrayList<>(filters);
            chain = new ArrayList<>();
            view.setAvailableFilters(availableFilters);
            view.setChain(chain);
        }));
    }

    public void initialize() {
        refreshRelay.call(null);
    }

    public void refresh() {
        refreshRelay.call(null);
    }

    public void addToChain(int selectedItemPosition) {
        if (move(availableFilters, chain, selectedItemPosition)) {
            view.setAvailableFilters(availableFilters);
            view.setChain(chain);
        }
    }

    public void removeFromChain(int selectedItemPosition) {
        if (move(chain, availableFilters, selectedItemPosition)) {
            view.setAvailableFilters(availableFilters);
            view.setChain(chain);
        }
    }

    static boolean move(List<Filter> from, List<Filter> to, int index) {
        if (index < 0 || index >= from.size()) {
            return false;
        }
        Filter selected = from.get(index);
        to.add(selected);
        from.remove(index);
        return true;
    }

    public void moveUpFilter(int index) {
        if (index < 1 || index >= chain.size()) {
            return;
        }
        Filter picked = chain.remove(index);
        chain.add(index - 1, picked);
        view.swapFilterInChain(index, index - 1);
    }

    public void moveDownFilter(int index) {
        if (index < 0 || index >= chain.size() - 1) {
            return;
        }
        Filter picked = chain.remove(index);
        chain.add(index + 1, picked);
        view.swapFilterInChain(index, index + 1);
    }

    public interface ChainComposerView extends PassiveView {
        void setAvailableFilters(List<Filter> filters);

        void setChain(List<Filter> chain);

        void showErrorMessage(String message);

        void swapFilterInChain(int from, int to);
    }
}

File: app/src/main/java/humbleactivity/app/ui/ChainComposer.kt
package humbleactivity.app.ui

import com.jakewharton.rxrelay2.PublishRelay
import humbleactivity.app.data.EffectorService
import humbleactivity.app.data.Filter
import humbleactivity.app.data.removeAt
import humbleactivity.app.data.swap
import io.reactivex.Observable
import javax.inject.Inject

class ChainComposer
@Inject
constructor(effectorService: EffectorService) {
    data class State(val availables: List<Filter>, val chain: List<Filter>) {
        fun refresh(availables: List<Filter>): State = State(availables, emptyList())

        fun addToChain(position: Int): State = copy(availables.removeAt(position), chain + availables[position])

        fun removeFromChain(position: Int): State = copy(availables + chain[position], chain.removeAt(position))

        fun moveDown(position: Int): State = copy(chain = chain.swap(position, position + 1))

        fun moveUp(position: Int): State = copy(chain = chain.swap(position, position - 1))
    }

    val _backdoor: PublishRelay<State> = PublishRelay.create()
    val refresh: PublishRelay<Unit> = PublishRelay.create()
    val addToChain: PublishRelay<Int> = PublishRelay.create()
    val removeFromChain: PublishRelay<Int> = PublishRelay.create()
    val moveUp: PublishRelay<Int> = PublishRelay.create()
    val moveDown: PublishRelay<Int> = PublishRelay.create()
    val loadError: PublishRelay<String> = PublishRelay.create()
    val chainCursor: PublishRelay<Int> = PublishRelay.create()
    val state: Observable<State> = Observable.mergeArray(
            _backdoor.map { newState -> { _: State -> newState } },
            refresh.concatMap {
                effectorService.listFilters()
                        .doOnError { throwable -> loadError.accept(throwable.message!!) }
                        .onErrorResumeNext(Observable.empty())
                        .map { response -> { state: State -> state.refresh(response) } }
            },
            addToChain.map { position -> { state: State -> state.addToChain(position) } },
            removeFromChain.map { position -> { state: State -> state.removeFromChain(position) } },
            moveUp.map { position ->
                { state: State ->
                    chainCursor.accept(position - 1);
                    state.moveUp(position)
                }
            },
            moveDown.map { position ->
                { state: State ->
                    chainCursor.accept(position + 1);
                    state.moveDown(position)
                }
            })
            .scan(State(emptyList(), emptyList()), { state: State, action: (State) -> State -> action(state) })
            .cacheWithInitialCapacity(1)
    val availables: Observable<List<Filter>> = state.map { it.availables }
    val chain: Observable<List<Filter>> = state.map { it.chain }

    fun initialize() {
        refresh.accept(Unit)
    }
}
